### PR TITLE
operator: Adjust CiliumEndpoint GC to only run once in kvstore mode

### LINKED ISSUE #24440: Cilium Operator fails to initialize when running with etcd with endpoint crd disabled
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

On a newly built cluster, with `disable-endpoint-crd: "true"` and etcd kvstore, the cilium-operator errors with 
```
level=warning msg="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: Failed to watch *v2.CiliumEndpoint: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=k8s
```

Expected below logs are never seen:
```
level=info msg="Starting CNP derivative handler" subsys=cilium-operator-generic
level=info msg="Starting CCNP derivative handler" subsys=cilium-operator-generic
level=info msg="Initialization complete" subsys=cilium-operator-generic
```

Note that this only happens on a cluster built directly with Cilium 1.11+. Upgrading from 1.10 (which creates the `ciliumendpoints.cilium.io` CRD), the issue is not seen.

### Cilium Version

1.11+ (I tested all 1.11.0 thru 1.11.15 as well as 1.12.1 and 1.13.1). Works fine with 1.10.18+

### Kernel Version

5.15.43-flatcar

### Kubernetes Version

v1.23.15

### Sysdump

_No response_

### Relevant log output

```shell
level=info msg="Cilium Operator 1.11.15 c5577e8 2023-03-15T19:19:16+01:00 go version go1.17.13 linux/amd64" subsys=cilium-operator-generic
level=info msg="Registering Operator metrics" subsys=metrics
level=info msg="Starting apiserver on address 127.0.0.1:9234" subsys=cilium-operator-api
level=info msg="Establishing connection to apiserver" host="https://192.168.0.1:443" subsys=k8s
level=info msg="Connected to apiserver" subsys=k8slevel=info msg="Connected to apiserver" subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumExternalWorkload/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumIdentity/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNode/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNetworkPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumClusterwideNetworkPolicy/v2 subsys=k8s
...
level=info msg="Starting to synchronize Kubernetes services to kvstore" subsys=watchers
level=info msg="Connected to the KVStore, syncing CiliumNodes to the KVStore" subsys=cilium-operator-generic
level=warning msg="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: Failed to watch *v2.CiliumEndpoint: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=k8s
level=warning msg="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/operator/watchers/cilium_endpoint.go:116: Failed to watch *v2.CiliumEndpoint: failed to list *v2.CiliumEndpoint: the server could not find the requested resource (get ciliumendpoints.cilium.io)" subsys=k8s
```

### Anything else?

For comparison, here are the operator v1.10.20 logs:
```
level=info msg="Cilium Operator 1.10.20 ad16558 2023-02-13T09:03:24+01:00 go version go1.16.15 linux/amd64" subsys=cilium-operator-generic
level=info msg="Registering Operator metrics" subsys=metrics
level=info msg="Starting apiserver on address 127.0.0.1:9234" subsys=cilium-operator-api
level=info msg="Establishing connection to apiserver" host="https://192.168.0.1:443" subsys=k8s
level=info msg="Connected to apiserver" subsys=k8s
level=info msg="Creating CRD (CustomResourceDefinition)..." name=CiliumEndpoint/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumEgressNATPolicy/v2alpha1 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumLocalRedirectPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumExternalWorkload/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNode/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumIdentity/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumClusterwideNetworkPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNetworkPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumEndpoint/v2 subsys=k8s
...
level=info msg="Starting to synchronize Kubernetes services to kvstore" subsys=watchers
level=info msg="Starting kvstore watchdog" interval=25s subsys=cilium-operator-generic
level=info msg="Removing Cilium Node Taints or Setting Cilium Is Up Condition for Kubernetes Nodes" k8sNamespace=kube-system label-selector="k8s-app=cilium" remove-cilium-node-taints=true set-cilium-is-up-condition=true subsys=cilium-operator-generic
level=info msg="Starting to garbage collect stale CiliumNode custom resources" subsys=watchers
level=info msg="Starting kvstore identity garbage collector" interval=15m0s subsys=cilium-operator-generic
level=info msg="Starting to garbage collect stale CiliumEndpoint custom resources" subsys=cilium-operator-generic
level=info msg="Garbage collected status/nodes in Cilium Clusterwide Network Policies found=0, gc=0" subsys=cilium-operator-generic
level=info msg="Garbage collected status/nodes in Cilium Network Policies found=5, gc=0" subsys=cilium-operator-generic
level=info msg="Starting CNP derivative handler" subsys=cilium-operator-generic
level=info msg="Starting CCNP derivative handler" subsys=cilium-operator-generic
level=info msg="Initialization complete" subsys=cilium-operator-generic
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by learnitall**
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Thanks for contributing!

This commit adjusts the logic that the operator uses to control endpoint garbage collection and syncing in order to account for the case when CiliumEndpoint CRDs are disabled in kvstore mode. The operator will now only start the garbage collector if CiliumEndpoint CRD mode is enabled, and if it isn't enabled, the operator will check to ensure that the CiliumEndpoint CRD is installed in the cluster before starting a one-off gc sync. This will allow users to use kvstore mode without having to worry about setting the endpoint gc interval to zero, and it covers the case where a user transitions from CRD mode to kvstore mode.

Fixes: cilium/cilium#24440

This issue is present in 1.11, 1.12, and 1.13. The 'if' block the commit adds to check the value of `DisableCiliumEndpointCRD` should be able to stay the same between versions, but its location will be in `operator/main.go` in versions 1.11 and 1.12.

```release-note
operator: Adjust CiliumEndpoint gc to account for kvstore mode
```

The following labels are applicable:

* `release-note/bug`
* `kind/bug`
* `needs-backport/1.11`
* `needs-backport/1.12`
* `needs-backport/1.13`

### CODE REVIEW

**Comment by learnitall** [comment]
From my understanding, we don't have a clear way to add operator tests. Can we merge this in even if it is not covered by an integration test so we can close out #24440 first and then look into testing?

**Review by squeed — APPROVED** [review]

**Review by aanm — CHANGES_REQUESTED** [review]
We can't disable the GC blindly because we might need to run it in the case a user disabled endpoint CRD after running Cilium with endpoint CRD enabled, meaning cilium endpoint customer resource will be left in the cluster without being GCed.

discussed offline

**Comment by learnitall** [comment]
I apologize for the delay in addressing the changes here- I got caught up in some other work. I changed the logic here to only enable long-term endpoint gc only if the `EndpointGCInterval` is non-zero _and_ if CiliumEndpiont CRDs are enabled. If these criteria aren't met, then I added a check to only run the one-off gc if CiliumEndpoint CRDs are installed in the cluster.

File: operator/cmd/root.go [hunk 1]
```diff
@@ -646,13 +647,26 @@ func (legacy *legacyOnLeader) onStart(_ hive.HookContext) error {
...
-		if operatorOption.Config.EndpointGCInterval != 0 {
+		if operatorOption.Config.EndpointGCInterval != 0 && !option.Config.DisableCiliumEndpointCRD {
 			enableCiliumEndpointSyncGC(legacy.ctx, &legacy.wg, legacy.clientset, ciliumNodeSynchronizer, false)
 		} else {
-			// Even if the EndpointGC is disabled we still want it to run at least
+			// Even if the EndpointGC or CEPs are disabled, we still want it to run at least
 			// once. This is to prevent leftover CEPs from populating ipcache with
 			// stale entries.
-			enableCiliumEndpointSyncGC(legacy.ctx, &legacy.wg, legacy.clientset, ciliumNodeSynchronizer, true)
+			// First check if CEP CRD is available, as it could be missing if CEPs are not enabled.
```
Comment (christarazi) [inline operator/cmd/root.go]: The context in the commit msg was very helpful to understand this code so could you elaborate here as well? I was initially confused until I went back and read your commit msg.

**Review by christarazi — APPROVED** [review]
LGTM pending on minor nit. AFAIU, this PR is not implementing CiliumEndpoint GC for kvstore, but rather making it work "better" when the user transitions to kvstore mode. If so, then makes sense. `enableCiliumEndpointSyncGC` looks like it only does K8s GC and does not GC from kvstore. Not sure if this is already known or not, or if I'm mistaken.

**Review by squeed — APPROVED** [review]

File: operator/cmd/root.go (hunk 1 above)
Comment (learnitall) [reply operator/cmd/root.go]: Ah for sure, thank you for bringing that up. I'll change my wording around to make it more clear and add the info from my commit message as a comment. From my understanding as well, this function doesn't do any GC for kvstore mode- this is taken care of by leases in etcd if I'm not mistaken.

Comment (christarazi) [reply operator/cmd/root.go]: Ah right--leases. Thanks!

**Comment by learnitall** [comment]
I expanded the comment to add more context and updated the PR/commit name to be more accurate. Should be good to go @christarazi

File: operator/cmd/root.go [hunk 2]
```diff
@@ -646,13 +647,29 @@ func (legacy *legacyOnLeader) onStart(_ hive.HookContext) error {
...
+			// Check if CEP CRD is available, as it could be missing if CEPs are not enabled.
+			// If the CRD is not available, then no garbage collection needs to be done.
+			_, err := legacy.clientset.ApiextensionsV1().CustomResourceDefinitions().Get(
+				legacy.ctx, k8sconstv2.CEPName, metav1.GetOptions{},
+			)
+
+			if err == nil {
+				enableCiliumEndpointSyncGC(legacy.ctx, &legacy.wg, legacy.clientset, ciliumNodeSynchronizer, true)
+			} else if k8sErrors.IsNotFound(err) {
+				log.WithError(err).Info("CiliumEndpoint CRD cannot be found, skipping garbage collector")
```
Comment (pippolo84) [inline operator/cmd/root.go]: typo: `collector` -> `collection`

**Review by pippolo84 — APPROVED** [review]
Just a typo, apart from that, LGTM!
Thanks! :100:

File: operator/cmd/root.go (hunk 2 above)
Comment (learnitall) [reply operator/cmd/root.go]: Fixed!

**Review by aanm — APPROVED** [review]

File: operator/cmd/root.go [hunk 3]
```diff
@@ -663,13 +664,29 @@ func (legacy *legacyOnLeader) onStart(_ hive.HookContext) error {
...
 			enableCiliumEndpointSyncGC(legacy.ctx, &legacy.wg, legacy.clientset, ciliumNodeSynchronizer, false)
 		} else {
-			// Even if the EndpointGC is disabled we still want it to run at least
-			// once. This is to prevent leftover CEPs from populating ipcache with
-			// stale entries.
-			enableCiliumEndpointSyncGC(legacy.ctx, &legacy.wg, legacy.clientset, ciliumNodeSynchronizer, true)
+			// Check if CEP CRD is available, as it could be missing if CEPs are not enabled.
+			// If the CRD is not available, then no garbage collection needs to be done.
+			_, err := legacy.clientset.ApiextensionsV1().CustomResourceDefinitions().Get(
+				legacy.ctx, k8sconstv2.CEPName, metav1.GetOptions{},
```
Comment (aanm) [inline operator/cmd/root.go]: Sorry, one minor nit. Similar to what we are doing in https://github.com/cilium/cilium/blob/8f9af4663ac75d5fa2e79200bd66625bf12b8639/pkg/k8s/watchers/endpointsynchronizer.go#L144-L147

We should do a get `meta_v1.GetOptions{ResourceVersion: "0"})`

**Review by aanm — CHANGES_REQUESTED** [review]

File: operator/cmd/root.go (hunk 3 above)
Comment (learnitall) [reply operator/cmd/root.go]: Good catch, fixed

**Review by aanm — APPROVED** [review]

**Comment by julianwiedmann** [comment]
> Failing test hit a known flake: #26669. All other tests passed and have 4/4 approved reviews, marking ready to merge.

I don't see why we're skipping the full test suite here. So let's run it.

**Comment by learnitall** [comment]
Ah my mistake, thanks for catching that. After the run you trigger @julianwiedmann , it looks like I hit two flakes:

1. https://github.com/cilium/cilium/issues/26669 (again): https://github.com/cilium/cilium/actions/runs/5546175386/jobs/10126187798?pr=25324
2. https://github.com/cilium/cilium-cli/issues/1794: https://github.com/cilium/cilium/actions/runs/5550923641/jobs/10136503171

**Comment by learnitall** [comment]
Hit https://github.com/cilium/cilium/issues/26669 again in https://github.com/cilium/cilium/actions/runs/5646165553/job/15293626629?pr=25324. Kicking off full test suite.

**Comment by learnitall** [comment]
Hit two new flakes on this last run:

* https://github.com/cilium/cilium/issues/27053
* https://github.com/cilium/cilium/issues/27052

**Pull request merged**